x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java

```diff
@@ -142,71 +142,72 @@ protected void doNextBulk(BulkRequest request, ActionListener<BulkResponse> next
-
-                    // This calls AsyncTwoPhaseIndexer#finishWithIndexingFailure
-                    // Determine whether the failure is irrecoverable (transform should go into failed state) or not (transform increments
-                    // the indexing failure counter
-                    // and possibly retries)
-                    Throwable irrecoverableException = ExceptionRootCauseFinder.getFirstIrrecoverableExceptionFromBulkResponses(
-                        deduplicatedFailures.values()
-                    );
-                    if (irrecoverableException == null) {
-                        String failureMessage = getBulkIndexDetailedFailureMessage("Significant failures: ", deduplicatedFailures);
-                        logger.debug("[{}] Bulk index experienced [{}] failures. {}", getJobId(), failureCount, failureMessage);
-
-                        Exception firstException = deduplicatedFailures.values().iterator().next().getFailure().getCause();
-                        nextPhase.onFailure(
-                            new BulkIndexingException(
-                                "Bulk index experienced [{}] failures. {}",
-                                firstException,
-                                false,
-                                failureCount,
-                                failureMessage
-                            )
-                        );
-                    } else {
-                        deduplicatedFailures.remove(irrecoverableException.getClass().getSimpleName());
-                        String failureMessage = getBulkIndexDetailedFailureMessage("Other failures: ", deduplicatedFailures);
-                        irrecoverableException = decorateBulkIndexException(irrecoverableException);
+            ActionListener.wrap(bulkResponse -> handleBulkResponse(bulkResponse, nextPhase), nextPhase::onFailure)
+        );
+    }
 
-                        logger.debug(
-                            "[{}] Bulk index experienced [{}] failures and at least 1 irrecoverable [{}]. {}",
-                            getJobId(),
-                            failureCount,
-                            ExceptionRootCauseFinder.getDetailedMessage(irrecoverableException),
-                            failureMessage
-                        );
+    protected void handleBulkResponse(BulkResponse bulkResponse, ActionListener<BulkResponse> nextPhase) {
+        if (bulkResponse.hasFailures() == false) {
+            // We don't know the of failures that have occurred (searching, processing, indexing, etc.),
+            // but if we search, process and bulk index then we have
+            // successfully processed an entire page of the transform and should reset the counter, even if we are in the middle
+            // of a checkpoint
+            context.resetReasonAndFailureCounter();
+            nextPhase.onResponse(bulkResponse);
+            return;
+        }
```

Comment:
OK, this diff is a pain. The only difference here is the `resetReasonAndFailureCounter` and the comments. I created a new method here so that this is testable.